bookkeeper-server/src/main/java/org/apache/bookkeeper/util/ByteBufVisitor.java

```diff
@@ -0,0 +1,1132 @@
+        @Override
+        protected byte[] initialValue() {
+            return new byte[TL_COPY_BUFFER_SIZE];
+        }
+    };
+
+    private static class InternalContext<T> {
+        int depth;
+        int maxDepth;
+        ByteBuf parentBuffer;
+        int parentOffset;
+        int parentLength;
+        T callbackContext;
+        ByteBufVisitorCallback<T> callback;
+        GetBytesCallbackByteBuf<T> callbackByteBuf = new GetBytesCallbackByteBuf(this);
+
+        void recursivelyVisitBuffers(ByteBuf visitBuffer, int visitIndex, int visitLength) {
+            // visit the wrapped buffers recursively if the buffer is not backed by an array or memory address
+            // and the max depth has not been reached
+            if (depth < maxDepth && !visitBuffer.hasMemoryAddress() && !visitBuffer.hasArray()) {
+                parentBuffer = visitBuffer;
+                parentOffset = visitIndex;
+                parentLength = visitLength;
+                depth++;
+                // call getBytes to trigger the wrapped buffer visit
+                visitBuffer.getBytes(visitIndex, callbackByteBuf, 0, visitLength);
+                depth--;
+            } else {
+                passBufferToCallback(visitBuffer, visitIndex, visitLength);
+            }
+        }
+
+        void handleBuffer(ByteBuf visitBuffer, int visitIndex, int visitLength) {
+            if (visitLength == 0) {
+                // skip visiting empty buffers
+                return;
+            }
+            if (visitBuffer == parentBuffer && visitIndex == parentOffset && visitLength == parentLength) {
+                // further recursion would cause unnecessary recursion up to the max depth of recursion
+                passBufferToCallback(visitBuffer, visitIndex, visitLength);
+            } else {
+                // use the doRecursivelyVisitBuffers method to visit the wrapped buffer, possibly recursively
+                recursivelyVisitBuffers(visitBuffer, visitIndex, visitLength);
+            }
+        }
+
+        private void passBufferToCallback(ByteBuf visitBuffer, int visitIndex, int visitLength) {
+            if (callback.preferArrayOrMemoryAddress(callbackContext)) {
+                if (visitBuffer.hasArray()) {
+                    handleArray(visitBuffer.array(), visitBuffer.arrayOffset() + visitIndex, visitLength);
+                } else if (visitBuffer.hasMemoryAddress() && callback.acceptsMemoryAddress(callbackContext)) {
+                    callback.visitBuffer(callbackContext, visitBuffer, visitIndex, visitLength);
+                } else if (callback.acceptsMemoryAddress(callbackContext) && visitBuffer.isDirect()
+                        && visitBuffer.alloc().isDirectBufferPooled()) {
+                    // read-only buffers need to be copied before they can be directly accessed
+                    ByteBuf copyBuffer = visitBuffer.copy(visitIndex, visitLength);
```

Comment:
We will copy the ByteBuf here, will this path will high-frequency called?

Reply:
When it gets here, the source buffer (`visitBuffer`) will need to be copied in any case.
Since this checks that the source buffer is a direct buffer and that the callback accepts a memory address buffer (direct buffer), it's better to make a direct copy than to use the thread local copy buffer (byte[]) for copying.

The last resort is copying the source buffer using the byte[] copying. 

Netty won't allow visiting the backing direct buffer or backing array when the buffer is a read-only buffer. In all other cases, the ByteBufVisitor is able to visit all source buffers without copies. This is the most efficient solution that there is available.